aimmo-game/simulation/turn_manager.py

```diff
@@ -80,13 +84,18 @@ def run_concurrent_turn(self):
         [thread.join() for thread in threads]
 
         cells_to_clear = set()
-        with game_state_provider as game_state:
-            while not action_queue.empty():
-                action = action_queue.get()[1]
-                action.apply(game_state.world_map)
-                cells_to_clear.add(action.target_location)
 
-            for cell in cells_to_clear:
+        while not action_queue.empty():
+            action = action_queue.get()[1]
+            with game_state_provider as game_state:
+                if action.is_legal(game_state.world_map):
+                    action.apply(game_state.world_map)
+                else:
+                    action.reject()
+            cells_to_clear.add(action.target_location)
+
+        for cell in cells_to_clear:
+            with game_state_provider as game_state:
                 game_state.world_map.clear_cell_actions(cell)
```

Comment:
Think we should call reject() on cleared actions to register the relevant events.